0023-keymanager-secret-sharing.md

```diff
@@ -0,0 +1,694 @@
+  /// for authorized nodes.
+  fn bivariate_share(&self, churp_id: u8, round: u64);
+
+  /// Construct a full share from the verified bivariate shares obtained
+  /// from the dealers.
+  ///
+  /// Full secret:
+  ///      s(y) =  B(node_id,y) = \sum_g_{node_id}(y)
+  fn dealing(&self, churp_id: u8, round: u64);
+
+  /// Construct a reduced share from the switch data points obtained from
+  /// the nodes in the current committee.
+  ///
+  /// Reduced share:
+  ///      r(x) = B(x, node_id)
+  fn share_reduction(&self, churp_id: u8, round: u64);
+
+  /// Randomize reduced share using bivariate shares obtained from
+  /// the nodes in the new committee.
+  ///
+  /// Proactive reduced share:
+  ///      r'(x) = B(x, node_id) + \sum Q(x, node_id)
+  fn proactive_randomization(&self, churp_id: u8, round: u64);
+
+  /// Construct a full share from the switch data points obtained from 
+  /// the nodes in the new committee.
+  ///
+  /// Full share:
+  ///      s'(y) = B'(node_id,y)
+  fn full_share_distribution(&self, churp_id: u8, round: u64);
+
+  /// Return switch data point (integer) for the calling node.
+  /// 
+  /// The point is evaluation of the shared secret (polynomial)
+  /// at the given y value (caller id).
+  /// 
+  /// Switch point:
+  ///     P = g(caller_id) = B(node_id,caller_id)
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the polynomial needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn switch_point(&self, churp_id: u8, round: u64) -> Integer;
+
+  /// Return the key share for the given round.
+  ///
+  /// To construct a derived key, the caller is required to collect 
+  /// at least the threshold number of key shares and combine them
+  /// locally to construct the derived key.
+  ///
+  /// Key share:
+  ///     K = H(runtime_id, key_pair_id)^{r g(0)}
+  ///
+  ///
+  /// The caller should always provide a proof which can be independently
+  /// verified before key shares are released to the caller.
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the key share needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn key_share(&self, churp_id: u8, round: u64, runtime_id: Namespace, 
+    key_pair_id: KeyPairId, proof: Proof) -> Integer;
+}
+```
+
+Methods:
+
+- `Churp.Init`
+
+- `Churp.VerificationMatrix`
+
+- `Churp.BivariateShare`
+
+- `Churp.Dealing`
+
+- `Churp.ShareReduction`
+
+- `Churp.ProactiveRandomization`
+
+- `Churp.FullShareDistribution`
+
+- `Churp.SwitchPoint`
+
+- `Churp.KeyShare`
+
+### Key manager client
+
+```rust
+/// Key manager client for CHURP.
+pub trait Client {
+  /// Return the key for the given round.
+  ///
+  /// The key is combined from key shares obtained from the key manager nodes
+  /// in the current CHURP committee.
+  fn key(&self, churp_id: u8, round: u64, runtime_id: Namespace, key_pair_id: KeyPairId) -> Vec<u8>;
```

Comment:
Churp ID and key manager runtime ID uniquely define a CHURP instance, so both parameters are required. 

I ommitted the `proof` because authorization should be done before the `key` method is called and defined in another adr.

Reply:
But runtime ID would be fixed to the originating runtime ID? Why would you want to override this?